src/test/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientConcurrencyTest.java

```diff
@@ -0,0 +1,389 @@
+    getClientCaller3 = this.getClientStart(getClientCaller3);
+    getClientCaller1 = this.getClientStart(getClientCaller1);
+    getClientCaller3 = this.getClientStart(getClientCaller3);
+    this.closeClientStart(closeClientCaller3, getClientCaller3.getClient());
+    getClientCaller3 = this.getClientStart(getClientCaller3);
+    getClientCaller3 = this.getClientStart(getClientCaller3);
+    getClientCaller2 = this.getClientStart(getClientCaller2);
+    this.closeClientStart(closeClientCaller2, getClientCaller2.getClient());
+
+    // join all threads
+    this.getClientStop(this.getCallers);
+    this.closeClientStop(this.closeCallers);
+
+    // verification
+    int totalGetCount = this.getCallers.size();
+    int totalCloseCount = this.closeCallers.size();
+
+    assert totalGetCount > totalCloseCount;
+
+    // should create as many times as it closes if param is enabled
+    this.getClientStop(this.getCallers)
+        .forEach(client -> StreamingClientHandler.isClientValid(client));
+    Mockito.verify(
+            this.streamingClientHandler,
+            Mockito.times(this.enableClientOptimization ? totalCloseCount : totalGetCount))
+        .createClient(Mockito.anyMap());
+    Mockito.verify(this.streamingClientHandler, Mockito.times(totalCloseCount))
+        .closeClient(Mockito.any(SnowflakeStreamingIngestClient.class));
+  }
+
+  @Test
+  public void testGetClientConcurrency() throws Exception {
+    // setup three get runners
+    ProviderCaller getClientCaller1 =
+        new ProviderCaller(
+            "getClient1",
+            ProviderMethods.GET_CLIENT,
+            this.streamingClientProvider,
+            this.clientConfig1);
+    ProviderCaller getClientCaller2 =
+        new ProviderCaller(
+            "getClient2",
+            ProviderMethods.GET_CLIENT,
+            this.streamingClientProvider,
+            this.clientConfig2);
+    ProviderCaller getClientCaller3 =
+        new ProviderCaller(
+            "getClient3",
+            ProviderMethods.GET_CLIENT,
+            this.streamingClientProvider,
+            this.clientConfig3);
+
+    // try a bunch of random get calls at the same time
+    this.getClientStart(getClientCaller1);
+    this.getClientStart(getClientCaller2);
+    this.getClientStart(getClientCaller2);
+    this.getClientStart(getClientCaller3);
+    this.getClientStart(getClientCaller1);
+    this.getClientStart(getClientCaller2);
+    this.getClientStart(getClientCaller3);
+    this.getClientStart(getClientCaller1);
+    this.getClientStart(getClientCaller2);
+    this.getClientStart(getClientCaller3);
+
+    // should create just once if param is enabled
+    this.getClientStop(this.getCallers)
+        .forEach(client -> Assert.assertTrue(StreamingClientHandler.isClientValid(client)));
+    Mockito.verify(
+            this.streamingClientHandler,
+            Mockito.times(this.enableClientOptimization ? 1 : this.getCallers.size()))
+        .createClient(Mockito.anyMap());
+  }
```

Comment:
Here is how countdownlatch would work. (Just an example)


```suggestion
  @Test
    public void testGetClientConcurrency() throws Exception {
        CountDownLatch latch = new CountDownLatch(10);

        // Create and submit tasks
        List<Future<SnowflakeStreamingIngestClient>> futures = new ArrayList<>();
        for (int i = 0; i < 10; i++) {
            Future<SnowflakeStreamingIngestClient> future = executorService.submit(() -> {
                SnowflakeStreamingIngestClient client = streamingClientProvider.getClient(clientConfig1);
                latch.countDown();
                return client;
            });
            futures.add(future);
        }

        // Waits for all tasks to complete
        latch.await();

        // Verify that clients are valid
        for (Future<SnowflakeStreamingIngestClient> future : futures) {
            Assert.assertTrue(StreamingClientHandler.isClientValid(future.get()));
        }

        // Verify that createClient() was called the expected number of times
        Mockito.verify(streamingClientHandler, Mockito.times(enableClientOptimization ? 1 : 10))
            .createClient(....);
    }
```

You can repeat this for all other tests and this doesnt require manual handling of threads/joining etc. (Makes your life much easier)

Reply:
THAT IS VERY HELPFUL